Jpa map storage: failure to process changelogs when using CockroachDB

### Describe the bug

After the addition of new unique constraints for the attributes tables, the changelogs can't be processed anymore when using `CockroachDB` because of lack of support for `DEFERRABLE` and `INITIALLY DEFERRED` options.

The new constraint looks like this:

```
        <addUniqueConstraint tableName="kc_client_scope_attribute" columnNames="fk_root, name, value" deferrable="true" initiallyDeferred="true" />
```

When Keycloak attempts to create the DB schemas using `CockroachDB`, it fails with:

```
Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set META-INF/client-scopes/jpa-client-scopes-changelog-1.xml::client-scopes-1::keycloak:
     Reason: liquibase.exception.DatabaseException: ERROR: at or near "EOF": syntax error: unimplemented: this syntax
  Detail: source SQL:
ALTER TABLE public.kc_client_scope_attribute ADD UNIQUE (fk_root, name, value) DEFERRABLE INITIALLY DEFERRED
                                                                                                           ^
  Hint: You have attempted to use a feature that is not yet implemented. See: https://go.crdb.dev/issue-v/31632/v21.2 
```

After reading the issue linked in the error description, I saw that there's no indication that support for these options will be added in the near future, so we might need to split these constraints into a separate changeset targeting `PostgreSQL` only, leaving the constraint out in case `CockroachDB` is used.

### Version

main

### Expected behavior

The changesets should be processed successfully when using both `PostgreSQL` and `CockroachDB`.

### Actual behavior

Changesets can't be processed when using `CockroachDB` due to unsupported options in the unique constraint found in the attributes tables.

### How to Reproduce?

Follow the instructions at https://github.com/keycloak/keycloak/discussions/9596 to start Keycloak with `CockroachDB` - just attempt to start the server with a clean database.

### Anything else?

_No response_

--- a/model/map-jpa/src/main/resources/META-INF/client-scopes/jpa-client-scopes-changelog-1.xml
+++ b/model/map-jpa/src/main/resources/META-INF/client-scopes/jpa-client-scopes-changelog-1.xml
@@ -57,8 +57,6 @@ limitations under the License.
             <column name="name" type="VARCHAR(255)"/>
             <column name="value" type="text"/>
         </createTable>
-        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
-        <addUniqueConstraint tableName="kc_client_scope_attribute" columnNames="fk_root, name, value" deferrable="true" initiallyDeferred="true" />
         <createIndex tableName="kc_client_scope_attribute" indexName="client_scope_attr_fk_root">
             <column name="fk_root"/>
         </createIndex>
@@ -69,6 +67,13 @@ limitations under the License.
         <modifySql dbms="postgresql,cockroachdb">
             <replace replace="VALUE(255)" with="(value::varchar(250))"/>
         </modifySql>
+    </changeSet>
+
+    <changeSet author="keycloak" id="client-scopes-2" dbms="postgresql">
+        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
+        <!-- this will not work on cockroachdb as deferred indexes are not supported in version 22.1 yet, therefore, only run it on postgresql -->
+        <!-- see https://go.crdb.dev/issue-v/31632/v21.2 for the current status of the implementation -->
+        <addUniqueConstraint tableName="kc_client_scope_attribute" columnNames="fk_root, name, value" deferrable="true" initiallyDeferred="true" />
     </changeSet>
 
 </databaseChangeLog>

--- a/model/map-jpa/src/main/resources/META-INF/clients/jpa-clients-changelog-1.xml
+++ b/model/map-jpa/src/main/resources/META-INF/clients/jpa-clients-changelog-1.xml
@@ -64,8 +64,6 @@ limitations under the License.
             <column name="name" type="VARCHAR(255)"/>
             <column name="value" type="text"/>
         </createTable>
-        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
-        <addUniqueConstraint tableName="kc_client_attribute" columnNames="fk_root, name, value" deferrable="true" initiallyDeferred="true" />
         <createIndex tableName="kc_client_attribute" indexName="client_attr_fk_root">
             <column name="fk_root"/>
         </createIndex>
@@ -76,6 +74,13 @@ limitations under the License.
         <modifySql dbms="postgresql,cockroachdb">
             <replace replace="VALUE(255)" with="(value::varchar(250))"/>
         </modifySql>
+    </changeSet>
+
+    <changeSet author="keycloak" id="clients-2" dbms="postgresql">
+        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
+        <!-- this will not work on cockroachdb as deferred indexes are not supported in version 22.1 yet, therefore, only run it on postgresql -->
+        <!-- see https://go.crdb.dev/issue-v/31632/v21.2 for the current status of the implementation -->
+        <addUniqueConstraint tableName="kc_client_attribute" columnNames="fk_root, name, value" deferrable="true" initiallyDeferred="true" />
     </changeSet>
 
 </databaseChangeLog>

--- a/model/map-jpa/src/main/resources/META-INF/groups/jpa-groups-changelog-1.xml
+++ b/model/map-jpa/src/main/resources/META-INF/groups/jpa-groups-changelog-1.xml
@@ -64,8 +64,6 @@ limitations under the License.
             <column name="name" type="VARCHAR(255)"/>
             <column name="value" type="text"/>
         </createTable>
-        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
-        <addUniqueConstraint tableName="kc_group_attribute" columnNames="fk_root, name, value" deferrable="true" initiallyDeferred="true" />
         <createIndex tableName="kc_group_attribute" indexName="group_attr_fk_root">
             <column name="fk_root"/>
         </createIndex>
@@ -76,6 +74,13 @@ limitations under the License.
         <modifySql dbms="postgresql,cockroachdb">
             <replace replace="VALUE(255)" with="(value::varchar(250))"/>
         </modifySql>
+    </changeSet>
+
+    <changeSet author="keycloak" id="groups-2" dbms="postgresql">
+        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
+        <!-- this will not work on cockroachdb as deferred indexes are not supported in version 22.1 yet, therefore, only run it on postgresql -->
+        <!-- see https://go.crdb.dev/issue-v/31632/v21.2 for the current status of the implementation -->
+        <addUniqueConstraint tableName="kc_group_attribute" columnNames="fk_root, name, value" deferrable="true" initiallyDeferred="true" />
     </changeSet>
 
 </databaseChangeLog>

--- a/model/map-jpa/src/main/resources/META-INF/realms/jpa-realms-changelog-1.xml
+++ b/model/map-jpa/src/main/resources/META-INF/realms/jpa-realms-changelog-1.xml
@@ -83,8 +83,6 @@ limitations under the License.
             <column name="name" type="VARCHAR(255)"/>
             <column name="value" type="TEXT"/>
         </createTable>
-        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
-        <addUniqueConstraint tableName="kc_realm_attribute" columnNames="fk_root, name, value" deferrable="true" initiallyDeferred="true" />
         <createIndex tableName="kc_realm_attribute" indexName="realm_attr_fk_root">
             <column name="fk_root"/>
         </createIndex>
@@ -95,6 +93,13 @@ limitations under the License.
         <modifySql dbms="postgresql,cockroachdb">
             <replace replace="VALUE(255)" with="(value::varchar(250))"/>
         </modifySql>
+    </changeSet>
+
+    <changeSet author="keycloak" id="realms-2" dbms="postgresql">
+        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
+        <!-- this will not work on cockroachdb as deferred indexes are not supported in version 22.1 yet, therefore, only run it on postgresql -->
+        <!-- see https://go.crdb.dev/issue-v/31632/v21.2 for the current status of the implementation -->
+        <addUniqueConstraint tableName="kc_realm_attribute" columnNames="fk_root, name, value" deferrable="true" initiallyDeferred="true" />
     </changeSet>
 
 </databaseChangeLog>

--- a/model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml
+++ b/model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml
@@ -66,8 +66,6 @@ limitations under the License.
             <column name="name" type="VARCHAR(255)"/>
             <column name="value" type="text"/>
         </createTable>
-        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
-        <addUniqueConstraint tableName="kc_role_attribute" columnNames="fk_root, name, value" deferrable="true" initiallyDeferred="true" />
         <createIndex tableName="kc_role_attribute" indexName="role_attr_fk_root">
             <column name="fk_root"/>
         </createIndex>
@@ -78,6 +76,13 @@ limitations under the License.
         <modifySql dbms="postgresql,cockroachdb">
             <replace replace="VALUE(255)" with="(value::varchar(250))"/>
         </modifySql>
+    </changeSet>
+
+    <changeSet author="keycloak" id="roles-2" dbms="postgresql">
+        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
+        <!-- this will not work on cockroachdb as deferred indexes are not supported in version 22.1 yet, therefore, only run it on postgresql -->
+        <!-- see https://go.crdb.dev/issue-v/31632/v21.2 for the current status of the implementation -->
+        <addUniqueConstraint tableName="kc_role_attribute" columnNames="fk_root, name, value" deferrable="true" initiallyDeferred="true" />
     </changeSet>
 
 </databaseChangeLog>

